### PR TITLE
[new_tool_suite]: pycytominer set of tools

### PR DESCRIPTION
**Pull request by rmassei**
Hi!
This PR introduces a new tools suite for the Python tool kit [Pycytominer](https://github.com/cytomining/pycytominer).

**Why do we need this?**
The tool kit is offering a series of commodity functions to perform data processing for image-based profiling. Many of these steps cannot be easily reproduce with existing General Text Tools (i.e. Datamash, Compute)  in Galaxy. **PRO:** Since the tools are build around the [CLI](https://pycytominer.readthedocs.io/en/stable/walkthroughs/pycytominer_cli.html), a nice documentation on their usage is already present online.

**Which CLI commands are wrapped?**

All five described in the [documentation](https://pycytominer.readthedocs.io/en/stable/walkthroughs/pycytominer_cli.html).

- Aggregate
- Normalize
- Annotate
- Feature Select
- Consensus

**Short comment on the column selection in the XML**
In the command line of the tool, I added this step:

columns_headers=`head -n1 "$profiles" | cut -d, -f"$strata"`

strata is a parameter with type="data_column" and use_header_names="true"

In my opinion, this is a nice approach because it lets the user select columns directly from the tool interface, where the headers are displayed. This reduces typing errors. However, since the parameter is passed as an integer, it needs to be converted into the corresponding column name (text) which is why this step is necessary.

**Tool I/O**
~~Pycytominer accept CSV and parquet files as input and can output also in both file formats. For simplicity, I just limited the I/O to CSV, Anyway there are already tools in Galaxy for parquet conversion.~~

Pycytominer accept CSV and parquet files as input and can output also in both file formats. However, the tool accept TSV and Tabular files (together with CSV) to better adapt to the Galaxy environment. Conversion is performed trough the Python script [format_converter_csvtsv.py](https://github.com/rmassei/galaxy-image-analysis/blob/a9bc5e28e9c330543f7288635fe09cae72d1e489/tools/pycitominer/format_converter_csvtsv.py) called within the command execution section. CSV output is then converted to tabular.

**Tests overview**
All to tests are based on the results of a previous test and they are checking the presence of certain values, number of lines and columns in the output CSV/TSV. The initial test data (01_single_cells.csv and platemap.csv) was created following the pycytominer documentation. I tested both CSV and TSV files.

---

### Check-list for the contributor

**Mandatory checks:**

* [X] I have read the [CONTRIBUTING.md](https://github.com/BMCV/galaxy-image-analysis/blob/master/CONTRIBUTING.md) document (last updated: 2026/03/06).
* [X] License permits unrestricted use (educational + commercial).
* [X] I agree to license these and all my past contributions to the Galaxy Image Analysis codebase under the [MIT license](https://github.com/BMCV/galaxy-image-analysis/blob/master/LICENSE.md).

**If this PR adds or updates a tool or tool collection:**

* [X] This PR adds a new tool or tool collection.
* [ ] This PR updates an existing tool or tool collection.
* [ ] Tools added/updated by this PR comply with the Guidelines below (or explain why they do not).

### Guidelines for the contributor

<details>
<summary>
    This section is cited from the <a href="https://doi.org/10.37044/osf.io/w8dsz">Naming and Annotation Conventions for Tools in the Image Community in Galaxy</a>.
</summary>

<h4>Naming</h4>

Generally, the name of Galaxy tools in our community should be expressive and concise, while stating the purpose of the tool as precisely as possible. Consistency of the namings of Galaxy tools is important to ensure they can be found easily. To maintain consistency, we consider phrasing names as imperatives a good practice, such as "Analyze particles" or "Perform segmentation using watershed transformation". An acknowledged exception from this rule is the names of tool wrappers of major tool suites, where the name of a tool wrapper should be chosen identically to the module or function of the tool which is wrapped (e.g., "MaskImage" in CellProfiler).

<h4>Tool description</h4>

If a Galaxy tool is a thin tool wrapper (e.g, part of a major tool suite), then the name of the wrapped tool (and only the name of the wrapped tool, subsequent to the term "with" as in "with Bioformats") should be used as the description of the tool (further examples include "with CellProfiler", "with ImageJ2", "with ImageMagick", "with SpyBOAT", "with SuperDSM"). This ensures that the tool is found by typing the name of the wrapped tool into the "Search" field on the Galaxy interface. The tool description should be empty if a tool is either not part of a major tool suite, or the main functionality of the tool is implemented in the wrapper.

<h4>Annotations</h4>

We point out that there is a global list of precedential annotations with <a href="https://bio.tools">Bio.tools</a> identifiers (Ison et al., 2019) in Galaxy (see <a href="https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/tool_util/ontologies/biotools_mappings.tsv">mappings</a>), which may outweigh the annotations made in the XML specification of a Galaxy tool (and thus the annotations of a tool reported within the web interface of Galaxy might be divergent). However, since the precedential annotations are subject to possible changes and to avoid redundant work, we do not aim to reflect those in our specifications (those which we make in the XML specifications of Galaxy tools).

</details>

### CODE REVIEW

**Review by kostrykin — COMMENTED** [review]
> Linting tool /home/runner/work/galaxy-image-analysis/galaxy-image-analysis/tools/pycitominer/pycytominer_annotate.xml
>   Failed linting
>   .. CHECK (TestsNoValid): 2 test(s) found.
>   .. INFO (OutputsNumber): 1 outputs found.
>   .. INFO (InputsNum): Found 4 input parameters.
>   .. CHECK (HelpPresent): Tool contains help section.
>   .. CHECK (HelpValidRST): Help contains valid reStructuredText.
>   .. CHECK (ToolIDValid): Tool defines an id [pycytominer_annotate].
>   .. CHECK (ToolNameValid): Tool defines a name [Annotate].
>   .. CHECK (ToolProfileValid): Tool specifies profile version [23.0].
>   .. CHECK (ToolVersionValid): Tool defines a version [1.6.1+galaxy0].
>   .. INFO (CommandInfo): Tool contains a command.
>   .. CHECK (CitationsFound): Found 1 citations.
>   .. INFO: URL OK https://pycytominer.readthedocs.io/en/stable/index.html
>   .. WARNING: remote_repository_url may be incorrect: expected it to end with a significant part of '/home/runner/work/galaxy-image-analysis/galaxy-image-analysis/tools/pycitominer'

The directory name is misspelled: Can you please move `tools/pycitominer` to `tools/pycytominer`?

**Comment by kostrykin** [comment]
I see the motivation of naming the tools in the suite by mapping the names closely to the respective commands of Pycytominer. For most of the tools/commands, this nicely agrees with our convention for tool names (`Aggregate`, `Annotate`, `Normalize`). I'm a bit bothered though that this doesn't hold for the `Consensus` tool,

> To maintain consistency, we consider **phrasing names as imperatives** a good practice, such as "Analyze particles" or "Perform segmentation using watershed transformation".

which, on the other hand, falls under the "acknowledged exception" rule:

> An **acknowledged exception from this rule is the names of tool wrappers of major tool suites**, where the name of a tool wrapper should be chosen identically to the module or function of the tool which is wrapped (e.g., "MaskImage" in CellProfiler).

So while I do not oppose the names of the tools that you have chosen, I'm still wondering whether we might enhance consistency and tool findability. In fact, you have already chosen a name for the `Select features` tool that subtly differs from the name of the wrapped command (`feature_select`), we might find something similar for the `Consensus` tool.

The following suggestions are based on the [descriptions of the commands in the Pycytominer docs](https://pycytominer.readthedocs.io/en/stable/walkthroughs/pycytominer_cli.html). What you think?

| Command name | Your suggestion | Maybe a more descriptive, imperative tool name? |
| --- | --- | --- |
| `aggregate` | Aggregate | Aggregate readouts |
| `annotate` | Annotate | Annotate readouts with metadata |
| `normalize` | Normalize | Normalize readouts |
| `feature_select` | Select features | Select informative features |
| `consensus` | Consensus | Compute consensus profiles |

I think it's good to mention `readouts` / `features` in the names of tools since—as far as I understand—Pycytominer essentially is a toolkit for (post-)processing of _readouts_ (aka previously extracted/quantified _features_). So people will hit those tools when looking for the corresponding key words.

**Comment by rmassei** [comment]
Hej @kostrykin, I agree with the renaming to make it more descriptive. 
I wonder if it is better to be even more consistent and do not use _readouts_ at all (which in my opinion quite general term) but just _features_. It sounds to be more a definition used for cell painting.
So:

| Command name | Your suggestion | Maybe a more descriptive, imperative tool name? |
| --- | --- | --- |
| `aggregate` | Aggregate | Aggregate **features** |
| `annotate` | Annotate | Annotate **features** with metadata |
| `normalize` | Normalize | Normalize **features** |
| `feature_select` | Select features | Select informative **features** |
| `consensus` | Consensus | Compute consensus profiles |

**Comment by kostrykin** [comment]
> Command name	Your suggestion	Maybe a more descriptive, imperative tool name?
> `aggregate`	Aggregate	Aggregate **features**
> `annotate`	Annotate	Annotate **features** with metadata
> `normalize`	Normalize	Normalize **features**
> `feature_select`	Select features	Select informative **features**
> `consensus`	Consensus	Compute consensus profiles

This was exactly what I had come up with at first :) then I gave it another thought, the _features_ are actually the columns, and the _readouts_ are the values. So, for example, `Annotate features with metadata` sounds like the _columns_ were annotated with metadata, which is not the case. `Normalize features` and `Aggregate features` are a bit less ambiguous, but still, they might suggest that the _set of columns_ were subject to change.

**Comment by rmassei** [comment]
Still, using **just** readouts sounds to me too general.
What about (just examples):

- Aggregate _plate_ readouts
- Aggregate _phenotype_ readouts
- Aggregate _cellular_ readouts

basically just adding an attribute to give context.

or then the tool name become excessively specific or verbose?

File: tools/pycytominer/pycytominer_aggregate.xml [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+python '$__tool_directory__/format_converter_csvtsv.py'
+    -i 'pycytominer_output/output.csv'
+    -o '$output'
+    -c csv2tsv
+    ]]></command>
+    <inputs>
+    <param name="profiles" type="data" format="csv,tsv,tabular" label="Input cell-profiling table" help= "Table with features extracted from biological images"/>
+    <param name="strata" type="data_column" format="csv,tsv,tabular"
+        label="Aggregation columns"
+        help="Select column names to aggregate by (i.e. Metadata_Plate, Metadata_Well)"
```
Comment (kostrykin) [inline tools/pycytominer/pycytominer_aggregate.xml]: You actually mean `e.g.,` here, like "for example"? `i.e.` means "in other words"

```suggestion
        help="Select column names to aggregate by (e.g., Metadata_Plate, Metadata_Well)"
```

File: tools/pycytominer/pycytominer_aggregate.xml [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    --operation '$operation' &&
+
+## Convert the result back to TSV for the Galaxy dataset
+python '$__tool_directory__/format_converter_csvtsv.py'
+    -i 'pycytominer_output/output.csv'
+    -o '$output'
+    -c csv2tsv
+    ]]></command>
+    <inputs>
+    <param name="profiles" type="data" format="csv,tsv,tabular" label="Input cell-profiling table" help= "Table with features extracted from biological images"/>
```
Comment (kostrykin) [inline tools/pycytominer/pycytominer_aggregate.xml]:
```suggestion
    <param name="profiles" type="data" format="csv,tsv,tabular" label="Input cell-profiling table" help="Table with features extracted from biological images"/>
```

File: tools/pycytominer/pycytominer_annotate.xml [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+python '$__tool_directory__/format_converter_csvtsv.py'
+    -i 'pycytominer_output/output.csv'
+    -o '$output'
+    -c csv2tsv
+    ]]></command>
+    <inputs>
+        <param name="profiles" type="data"  format="csv,tsv,tabular" label="Input cell-profiling table" help= "Table with features extracted from biological images"/>
+        <param name="profiles_joiner" type="data_column"   format="csv,tsv,tabular" 
+            label="Column describing the wells"
+            help="Column with well values in the cell-profiling table (i.e. Metadata_Well)"
```
Comment (kostrykin) [inline tools/pycytominer/pycytominer_annotate.xml]: You actually mean `e.g.,` here, like "for example"? `i.e.` means "in other words"


```suggestion
            help="Column with well values in the cell-profiling table (e.g., Metadata_Well)"
```

File: tools/pycytominer/pycytominer_annotate.xml [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+--join_on "Metadata_\$columns_headers2, \$columns_headers1" &&
+
+## Convert the result back to TSV for the Galaxy dataset
+python '$__tool_directory__/format_converter_csvtsv.py'
+    -i 'pycytominer_output/output.csv'
+    -o '$output'
+    -c csv2tsv
+    ]]></command>
+    <inputs>
+        <param name="profiles" type="data"  format="csv,tsv,tabular" label="Input cell-profiling table" help= "Table with features extracted from biological images"/>
```
Comment (kostrykin) [inline tools/pycytominer/pycytominer_annotate.xml]:
```suggestion
        <param name="profiles" type="data" format="csv,tsv,tabular" label="Input cell-profiling table"help= "Table with features extracted from biological images"/>
```

File: tools/pycytominer/pycytominer_annotate.xml [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+        <param name="profiles_joiner" type="data_column"   format="csv,tsv,tabular"
```
Comment (kostrykin) [inline tools/pycytominer/pycytominer_annotate.xml]:
```suggestion
        <param name="profiles_joiner" type="data_column" format="csv,tsv,tabular" 
```

File: tools/pycytominer/pycytominer_annotate.xml [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+    -c csv2tsv
+    ]]></command>
+    <inputs>
+        <param name="profiles" type="data"  format="csv,tsv,tabular" label="Input cell-profiling table" help= "Table with features extracted from biological images"/>
+        <param name="profiles_joiner" type="data_column"   format="csv,tsv,tabular" 
+            label="Column describing the wells"
+            help="Column with well values in the cell-profiling table (i.e. Metadata_Well)"
+            data_ref="profiles" 
+            use_header_names="true" />
+        <param name="platemap" type="data"  format="csv,tsv,tabular" label="Input platemap table" help= "Table with platempa information such a treatment, celline and concentration tested"/>
```
Comment (kostrykin) [inline tools/pycytominer/pycytominer_annotate.xml]:
```suggestion
        <param name="platemap" type="data" format="csv,tsv,tabular" label="Input platemap table" help="Table with plate map information such a treatment, cell line, and concentration tested"/>
```

File: tools/pycytominer/pycytominer_consensus.xml [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+--replicate_columns "\$columns_headers" &&
+
+## Convert the result back to TSV for the Galaxy dataset
+python '$__tool_directory__/format_converter_csvtsv.py'
+    -i 'pycytominer_output/output.csv'
+    -o '$output'
+    -c csv2tsv
+    ]]></command>
+    <inputs>
+        <param name="profiles" type="data" format="csv,tsv,tabular" label="Input cell-profiling table" help= "Table with features extracted from biological images, normally after feature selection"/>
```
Comment (kostrykin) [inline tools/pycytominer/pycytominer_consensus.xml]:
```suggestion
        <param name="profiles" type="data" format="csv,tsv,tabular" label="Input cell-profiling table" help="Table with features extracted from biological images, normally after feature selection"/>
```

File: tools/pycytominer/pycytominer_feature.xml [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+--operation "${', '.join($operation)}" &&
+
+## Convert the result back to TSV for the Galaxy dataset
+python '$__tool_directory__/format_converter_csvtsv.py'
+    -i 'pycytominer_output/output.csv'
+    -o '$output'
+    -c csv2tsv
+    ]]></command>
+    <inputs>
+        <param name="profiles" type="data" format="csv,tsv,tabular" label="Input cell-profiling table" help= "Table with features extracted from biological images, normally after normalization"/>
```
Comment (kostrykin) [inline tools/pycytominer/pycytominer_feature.xml]:
```suggestion
        <param name="profiles" type="data" format="csv,tsv,tabular" label="Input cell-profiling table" help="Table with features extracted from biological images, normally after normalization"/>
```

File: tools/pycytominer/pycytominer_feature.xml [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+        <param name="operation"
+            label="Select operations"
+            help="List of operation to apply to the cell-profiling table"
+            type="select"
+            multiple="true">
```
Comment (kostrykin) [inline tools/pycytominer/pycytominer_feature.xml]: Currently this is marked as optional:

Image

Does this actually make sense?

I think, at least one must be selected?

Otherwise, can you maybe add a test that covers an empty input?

File: tools/pycytominer/pycytominer_normalize.xml [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+--method '$method' &&
+
+## Convert the result back to TSV for the Galaxy dataset
+python '$__tool_directory__/format_converter_csvtsv.py'
+    -i 'pycytominer_output/output.csv'
+    -o '$output'
+    -c csv2tsv
+    ]]></command>
+    <inputs>
+        <param name="profiles" type="data" format="csv,tsv,tabular" label="Input cell-profiling table" help= "Table with features extracted from biological images, normally after annotation with treatments"/>
```
Comment (kostrykin) [inline tools/pycytominer/pycytominer_normalize.xml]:
```suggestion
        <param name="profiles" type="data" format="csv,tsv,tabular" label="Input cell-profiling table" help="Table with features extracted from biological images, normally after annotation with treatments"/>
```

File: tools/pycytominer/pycytominer_normalize.xml [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+        <param name="column" type="data_column"  format="csv,tsv,tabular"
```
Comment (kostrykin) [inline tools/pycytominer/pycytominer_normalize.xml]:
```suggestion
        <param name="column" type="data_column" format="csv,tsv,tabular" 
```

File: tools/pycytominer/pycytominer_normalize.xml [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+python '$__tool_directory__/format_converter_csvtsv.py'
+    -i 'pycytominer_output/output.csv'
+    -o '$output'
+    -c csv2tsv
+    ]]></command>
+    <inputs>
+        <param name="profiles" type="data" format="csv,tsv,tabular" label="Input cell-profiling table" help= "Table with features extracted from biological images, normally after annotation with treatments"/>
+        <param name="column" type="data_column"  format="csv,tsv,tabular" 
+            label="Column with normalization values"
+            help="Column containing the values to use for normalized (i.e. Metadata_treatment)"
```
Comment (kostrykin) [inline tools/pycytominer/pycytominer_normalize.xml]: You actually mean `e.g.,` here, like "for example"? `i.e.` means "in other words"

```suggestion
            help="Column containing the values to use for normalized (e.g., Metadata_treatment)"
```

File: tools/pycytominer/pycytominer_normalize.xml [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+            data_ref="profiles" 
+            use_header_names="true"/>
+        <param name="treatment" type="text" label="Value" help="Wich value to use for normalized (i.e. DMSO, Control)"/>
```
Comment (kostrykin) [inline tools/pycytominer/pycytominer_normalize.xml]: Does this input support any values aside of `DMSO`, `Control`, and `all`? My impression from the documentation and the tests is that it is restricted to those three. In that case, it should be a `type="select"` field, no?

File: tools/pycytominer/pycytominer_annotate.xml [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+        <param name="platemap_joiner" type="data_column"   format="csv,tsv,tabular"
```
Comment (kostrykin) [inline tools/pycytominer/pycytominer_annotate.xml]:
```suggestion
        <param name="platemap_joiner" type="data_column" format="csv,tsv,tabular" 
```

File: tools/pycytominer/pycytominer_annotate.xml [hunk 15]
```diff
@@ -0,0 +1,97 @@
...
+    <inputs>
+        <param name="profiles" type="data"  format="csv,tsv,tabular" label="Input cell-profiling table" help= "Table with features extracted from biological images"/>
+        <param name="profiles_joiner" type="data_column"   format="csv,tsv,tabular" 
+            label="Column describing the wells"
+            help="Column with well values in the cell-profiling table (i.e. Metadata_Well)"
+            data_ref="profiles" 
+            use_header_names="true" />
+        <param name="platemap" type="data"  format="csv,tsv,tabular" label="Input platemap table" help= "Table with platempa information such a treatment, celline and concentration tested"/>
+        <param name="platemap_joiner" type="data_column"   format="csv,tsv,tabular" 
+            label="Column describing the wells"
```
Comment (kostrykin) [inline tools/pycytominer/pycytominer_annotate.xml]: The inputs `profiles_joiner` and `platemap_joiner` have identical labels. This is not so nice, because it creates ambiguities when one wants to point someone to a specific field (e.g., when writing tutorials).

File: tools/pycytominer/pycytominer_feature.xml (hunk 9 above)
Comment (rmassei) [reply tools/pycytominer/pycytominer_feature.xml]: totally right, this value is not optional at all, will correct it!

File: tools/pycytominer/pycytominer_normalize.xml (hunk 13 above)
Comment (rmassei) [reply tools/pycytominer/pycytominer_normalize.xml]: mmm no, I am quite sure that it supports any string value in the "Metadata_treatment" column.

File: tools/pycytominer/pycytominer_annotate.xml (hunk 15 above)
Comment (rmassei) [reply tools/pycytominer/pycytominer_annotate.xml]: true, will correct it!

**Comment by rmassei** [comment]
I also corrected all labels of the `profiles` input in all the XML

before:
`label="Input cell-profiling table"`
now:
`label="Input feature-readouts table"`

This is to generalize more the tool input.

File: tools/pycytominer/.shed.yml [hunk 16]
```diff
@@ -0,0 +1,33 @@
+owner: ufz
```
Comment (kostrykin) [inline tools/pycytominer/.shed.yml]: See this: https://github.com/BMCV/galaxy-image-analysis/pull/207/changes#diff-8ef3f15c81929892795d0d192abb750b4d069dd4f7cb503ddcc064f053fb0e12
```suggestion
owner: imgteam
```

This is required unless `ufz` is a recognized owner on the Galaxy ToolShed, but I think it isn't: https://toolshed.g2.bx.psu.edu/repositories_by_owner

Comment (rmassei) [reply tools/pycytominer/.shed.yml]: ah, yes, I remember this issue also with a previous tool, no problem to change the owner

**Review by kostrykin — APPROVED** [review]
Ready to merge? @rmassei

**Comment by rmassei** [comment]
Ready! Thanks a lot for all the review and inputs!

**Comment by kostrykin** [comment]
Thank you! @rmassei 

Don't forget to add the tools on EU :)

**Comment by kostrykin** [comment]
Attention: deployment skipped!

https://github.com/BMCV/galaxy-image-analysis/actions/runs/32248025500

**Comment by kostrykin** [comment]
Attention: deployment skipped!

https://github.com/BMCV/galaxy-image-analysis/actions/runs/32248025500

**Comment by kostrykin** [comment]
Attention: deployment skipped!

https://github.com/BMCV/galaxy-image-analysis/actions/runs/32248025500

**Comment by kostrykin** [comment]
I will try to update the CI. I will let you know when the deployment should be fixed. @rmassei

**Comment by kostrykin** [comment]
Attention: deployment cancelled!

https://github.com/BMCV/galaxy-image-analysis/actions/runs/32248025500
